drgn/helpers/linux/runq.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2025, Oracle and/or its affiliates.
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""
+runq
+-------------
+
+The ``drgn.helpers.linux.runq`` module provides helpers for working with the
+Linux runq.
+"""
+from drgn import Object, Program
+from drgn.helpers.linux.percpu import per_cpu
+from drgn.helpers.linux.sched import cpu_curr, task_cpu
+
+
+def runq_clock(prog: Program, cpu: int) -> int:
+    """
+    Get clock of cpu runqueue ``struct rq``
+
+    :param prog: drgn program
+    :param cpu: cpu index
+    :returns: cpu runqueue clock in ns granularity
+    """
+    rq = per_cpu(prog["runqueues"], cpu)
+    return rq.clock.value_()
+
+
+def get_task_arrival_time(task: Object) -> int:
+    """
+    Get a task's arrival time on cpu
+
+    A task's arrival time is only updated when the task is put ON a cpu via
+    context_switch.
+
+    :param task: ``struct task_struct *``
+    :returns: arrival time instance in ns granularity
+    """
+
+    try:
+        arrival_time = task.last_run.value_()
+    except AttributeError:
+        try:
+            arrival_time = task.timestamp.value_()
+        except AttributeError:
+            arrival_time = task.sched_info.last_arrival.value_()
+
+    return arrival_time
+
+
+def task_lastrun2now(task: Object) -> int:
+    """
+    Get the duration from task last run timestamp to now
+
+    The return duration will cover task's last run time on cpu and also
+    the time staying in current status, usually the time slice for task
+    on cpu will be short, so this can roughly tell how long this task
+    has been staying in current status.
+    For task status in "RU" status, if it's still on cpu, then this return
+    the duration time this task has been running, otherwise it roughly tell
+    how long this task has been staying in runqueue.
+
+    :param prog: drgn program
+    :param task: ``struct task_struct *``
+    :returns: duration in ns granularity
+    """
+    prog = task.prog_
+    arrival_time = get_task_arrival_time(task)
+    rq_clock = runq_clock(prog, task_cpu(task))
+
+    return rq_clock - arrival_time
```

Comment:
This needs a unit test. Take a look at the function `fork_and_stop()` in `tests/linux_kernel/__init__.py` - you can use that to create a task in the live kernel, and you could use this helper and at least assert that the result is above or below certain thresholds.